src/connector/src/sink/snowflake_connector.rs

```diff
@@ -0,0 +1,240 @@
+        })?;
+        Ok(jwt_token)
+    }
+
+    fn build_request_and_client(&self) -> (Builder, Client<HttpsConnector<HttpConnector>>) {
+        let builder = Request::post(self.url.clone());
+
+        let connector = HttpsConnector::new();
+        let client = Client::builder()
+            .pool_idle_timeout(POOL_IDLE_TIMEOUT)
+            .build(connector);
+
+        (builder, client)
+    }
+
+    /// NOTE: this function should ONLY be called *after*
+    /// uploading files to remote external staged storage, i.e., AWS S3
+    pub async fn send_request(&self, file_num: u32) -> Result<()> {
+        let (builder, client) = self.build_request_and_client();
+
+        // Generate the jwt_token
+        let jwt_token = self.generate_jwt_token()?;
+        let builder = builder
+            .header(header::CONTENT_TYPE, "text/plain")
+            .header("Authorization", format!("Bearer {}", jwt_token))
+            .header(
+                "X-Snowflake-Authorization-Token-Type".to_string(),
+                "KEYPAIR_JWT",
+            );
+
+        let request = builder
+            .body(Body::from(format!(
+                "{}_{}",
+                S3_INTERMEDIATE_FILE_NAME, file_num
+            )))
+            .map_err(|err| SinkError::Snowflake(err.to_string()))?;
+
+        let response = client
+            .request(request)
+            .await
+            .map_err(|err| SinkError::Snowflake(err.to_string()))?;
+
+        if response.status() != StatusCode::OK {
+            return Err(SinkError::Snowflake(format!(
+                "failed to make http request, error code: {}\ndetailed response: {:#?}",
+                response.status(),
+                response,
+            )));
+        }
+
+        Ok(())
+    }
+}
+
+/// TODO(Zihao): refactor this part after s3 sink is available
+pub struct SnowflakeS3Client {
+    s3_bucket: String,
+    s3_client: S3Client,
+}
+
+impl SnowflakeS3Client {
+    pub async fn new(
+        s3_bucket: String,
+        aws_access_key_id: String,
+        aws_secret_access_key: String,
+        aws_region: String,
+    ) -> Self {
+        let credentials = Credentials::new(
+            aws_access_key_id,
+            aws_secret_access_key,
+            // we don't allow temporary credentials
+            None,
+            None,
+            "rw_sink_to_s3_credentials",
+        );
+
+        let region = RegionProviderChain::first_try(Region::new(aws_region)).or_default_provider();
+
+        let config = aws_config::from_env()
+            .credentials_provider(credentials)
+            .region(region)
+            .load()
+            .await;
+
+        // create the brand new s3 client used to sink files to s3
+        let s3_client = S3Client::new(&config);
+
+        Self {
+            s3_bucket,
+            s3_client,
+        }
+    }
+
+    pub async fn sink_to_s3(&self, data: Bytes, file_num: u32) -> Result<()> {
```

Comment:
Is it possible to use S3 multi-part upload here? As far as I know, multi-part upload can help us upload big file (good for throughput) while using little memory buffer, which looks like a perfect match for our use case.

Reply:
after some investigations, here are my concerns:

1. for the use-case of our snowflake sink, the size of batched payload (i.e., the rows) typically should not be set as too large - reason is the data freshness / visibility; if user set the rows to batch to, e.g., 100000 or more, which could potentially utilize multi-part upload (since the size of payload would generally be > 100MB), however then it will take **much longer time** for the data to be batched and sink to downstream (a.k.a. visible to the users) - which I think most of the users may not want to see this.

2. according to official document, it's recommmend to use multi-part upload if the size of the object to be uploaded exceeds 100MB (as previously mentioned) - which should not be a common case for the use-case of snowflake sink; especially the sink is a two-step approach, sinking from s3 to snowflake pipe also takes time.

so my preference is to keep it as simple as possible at present, the current approach will ensure data freshness, plus the flexibility for users to adjust batch size in a reasonable range.